nrf_wifi/hw_if/hal/src/hal_interrupt.c

```diff
@@ -480,8 +497,55 @@ static unsigned int hal_rpu_event_get_all(struct nrf_wifi_hal_dev_ctx *hal_dev_c
 	return num_events;
 }
 
+static enum nrf_wifi_status hal_rpu_process_wdog(struct nrf_wifi_hal_dev_ctx *hal_dev_ctx,
+						  bool *do_rpu_recovery)
+{
+	enum nrf_wifi_status nrf_wifi_status = NRF_WIFI_STATUS_FAIL;
+	bool rpu_recovery = false;
+#ifdef CONFIG_NRF_WIFI_LOW_POWER
+	unsigned long flags = 0;
+#endif
+
+	nrf_wifi_osal_log_dbg(hal_dev_ctx->hpriv->opriv,
+			      "Processing watchdog interrupt");
+
+	/* Check if host has asserted WAKEUP_NOW */
+#ifdef CONFIG_NRF_WIFI_LOW_POWER
+	nrf_wifi_osal_spinlock_irq_take(hal_dev_ctx->hpriv->opriv,
+					hal_dev_ctx->rpu_ps_lock,
+					&flags);
+
+	if (hal_dev_ctx->rpu_ps_state == RPU_PS_STATE_AWAKE) {
+		nrf_wifi_osal_spinlock_irq_rel(hal_dev_ctx->hpriv->opriv,
+					       hal_dev_ctx->rpu_ps_lock,
+					       &flags);
+		nrf_wifi_osal_log_dbg(hal_dev_ctx->hpriv->opriv,
+				      "Host has asserted WAKEUP_NOW, ignoring watchdog interrupt");
+		goto out;
+	}
+
+	rpu_recovery = true;
+	nrf_wifi_osal_spinlock_irq_rel(hal_dev_ctx->hpriv->opriv,
+				       hal_dev_ctx->rpu_ps_lock,
+				       &flags);
+#endif /* CONFIG_NRF_WIFI_LOW_POWER */
 
-enum nrf_wifi_status hal_rpu_irq_process(struct nrf_wifi_hal_dev_ctx *hal_dev_ctx)
+	if (!rpu_recovery) {
+		hal_rpu_irq_wdog_rearm(hal_dev_ctx);
+		goto out;
+	}
+
+	nrf_wifi_osal_log_dbg(hal_dev_ctx->hpriv->opriv,
+			      "Host has not asserted WAKEUP_NOW, start RPU recovery");
+out:
+	nrf_wifi_status = NRF_WIFI_STATUS_SUCCESS;
```

Comment:
Why do we need `nrf_wifi_status` if we are always going to return `SUCCESS` ?

Reply:
To avoid API changes at all layers, I had defined this function that returns status, there is no usecase for that now, but might come in handy later and we can just update the status.